gitpod-io/gitpod
# Fix repo names for Bitbucket (Server)

## Description

I was looking into ENT-724 and stumbled upon a bug with Bitbucket Server: when we resolve the context of a BBS URL, we take the `.name` from the response and use that for subsequent API requests as part of the URL (fetch `/{user}/{name}/resource`). The problem is that `.name` is actually the human-readable label for repositories and not the URL part, so we needed to swap it with `.slug` to prevent these issues from happening.

Additionally, when I was already changing names to slugs, I figured we could use the display name when importing repositories. Thus, I introduced `displayName` on the general `Repository` interface to let compatible providers (BB and BBS AFAIK) feed us with something more human-readable for our default configuration names.

Furthermore, I removed `slug` from `CreateProjectParams` as we didn't use it anywhere and don't expose that to anyone.

Fixes ENT-850

## How to test


Import the following repos and check out their default names:
- https://bitbucket.org/test-gitpod-ext/wassup-homies/src/main/ (should be `Wassup homies`)
- https://bitbucket.gitpod-dev.com/users/filip/repos/spring-petclinic/browse (should be `Spring Petclinic`)

Check out the tests for a conclusive update in behavior. You need a bunch of setup and config for tests to work, so these are the results I can share when running against b7c832fdf41920ae2496fbc3921e01ee6e1b1905:

```
gitpod /workspace/gitpod/components/server (ft/use-correct-name-with-bbs) $ yarn build && yarn mocha './**/bitbucket*-context-parser.spec.js' --
exclude './node_modules/**' --exit
yarn run v1.22.22
$ yarn generate && npx tsc
$ leeway run components/spicedb:generate-ts > src/authorization/definitions.ts && npx prettier --write src/authorization/definitions.ts
src/authorization/definitions.ts 172ms
Done in 8.61s.
yarn run v1.22.22
$ /workspace/gitpod/node_modules/.bin/mocha './**/bitbucket*-context-parser.spec.js' --exclude './node_modules/**' --exit


  TestBitbucketServerContextParser
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo/default-branch - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo/branches?filterText=master&boostMatches=true - OK
    ✔ test_tree_context_01 (153ms)
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/git/repos/gitpod-test-repo - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/git/repos/gitpod-test-repo/default-branch - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/git/repos/gitpod-test-repo/branches?filterText=master&boostMatches=true - OK
    ✔ test_tree_context_02 (117ms)
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/users/geropl/repos/test-user-repo - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/users/geropl/repos/test-user-repo/default-branch - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/users/geropl/repos/test-user-repo/branches?filterText=main&boostMatches=true - OK
    ✔ test_tree_context_03_user_repo (118ms)
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/users/geropl/repos/test-user-repo/default-branch - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/users/geropl/repos/test-user-repo - OK
    ✔ test_commit_context_01_user_repo (68ms)
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo/default-branch - OK
    ✔ test_commit_context_02 (67ms)
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo/default-branch - OK
    ✔ test_commit_context_03_with_branch_ref (69ms)
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/users/svenefftinge/repos/browser-extension-test - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/users/svenefftinge/repos/browser-extension-test/default-branch - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/users/svenefftinge/repos/browser-extension-test/branches?filterText=my-branch&boostMatches=true - OK
    ✔ test_branch_context_01 (123ms)
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/users/svenefftinge/repos/browser-extension-test/branches?details=true&filterText=my-branch&boostMatches=true - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/users/svenefftinge/repos/browser-extension-test - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/users/svenefftinge/repos/browser-extension-test/default-branch - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/users/svenefftinge/repos/browser-extension-test/branches?filterText=my-branch&boostMatches=true - OK
    ✔ test_branch_context_02 (191ms)
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo/pull-requests/1 - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo/default-branch - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo/default-branch - OK
    ✔ test_PR_context_01 (158ms)
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo/pull-requests/1 - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo/default-branch - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo/default-branch - OK
    ✔ test_PR_context_02 (180ms)
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo/default-branch - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo/tags?filterText=test-tag-v1.0.1 - OK
    ✔ test_tag_context_01 (123ms)
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo/branches?details=true&filterText=test-tag-v1.0.1&boostMatches=true - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo/tags?filterText=test-tag-v1.0.1 - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo/default-branch - OK
BBS: GET https://bitbucket.gitpod-dev.com/rest/api/1.0/projects/GIT/repos/gitpod-test-repo/tags?filterText=test-tag-v1.0.1 - OK
    ✔ test_tag_context_02 (222ms)
    ✔ test_toSimpleBranchName

  TestBitbucketContextParser
    ✔ testCanHandleBitbucketRepo
    ✔ testCanNotHandleGitHubRepo
    ✔ testShortContext_01 (667ms)
    ✔ testSrcContext_01 (889ms)
    ✔ testSrcContext_02 (500ms)
    ✔ testSrcContext_03 (841ms)
    ✔ testCommitsContext_01 (660ms)
    ✔ testBranchContext_01 (619ms)
    ✔ testBranchContext_02 (1078ms)
    ✔ testBranchContext_03 (654ms)
    ✔ testCommitsContext_02 (695ms)
    ✔ testSrcContext_04 (820ms)
    ✔ testCommitsContext_03 (233ms)
    ✔ testFileContext_01 (1113ms)
    ✔ testFileContext_02 (1140ms)
    ✔ testDirectoryContext_01 (1139ms)
    ✔ testPullRequestContext_01 (779ms)
    ✔ testPullRequestContext_02 (1057ms)
    ✔ testIssueContext_01 (847ms)
    ✔ testSrcContextForkedRepo_01 (1083ms)


  33 passing (16s)
```
